### ISSUE TITLE
Should `BlobId` type be aligned with `attachments` type in `Observation` schema?

### ISSUE THREAD

**Issue by achou11** [digidem/mapeo-core-next#346]
## Description

Currently running into a slight mismatch in types and not entirely sure where a change needs to be done:

`BlobId`, which is returned by our BlobApi when creating what are eventually observation attachments:

```ts
type BlobId = {
  type: 'photo' | 'audio' | 'video'
  variant: ... // omitted for brevity
  name: string
  driveId: string
}
```

`Attachment`, defined by `@mapeo/schema` in the Observation schema:

```ts
type Attachment = {
  type: 'photo' | 'audio' | 'video' | 'UNRECOGNIZED'
  hash: string
  name: string
  driveDiscoveryId: string
}
```

The primary mismatch is the `type` field for each. My guess is that `@mapeo/core` should be updated to align with schema

**Comment by gmaclennan** [comment]
Thoughts:
- `type` is correct as-is: an observation that is read from a hypercore could have an attachment type that is `UNRECOGNIZED` - it could have been created by a future version of CoMapeo and this version does not know about it yet. However the blob API should not allow us to create a blob of type `UNRECOGNIZED`, and should not allow reading it - it is the responsibility of the API consumer to ignore attachments that are `UNRECOGNIZED` (because we don't know how to display them - they could be any kind of file, e.g. sensor readings, 3-d models etc.
- `driveId` for the BlobId should be changed to `driveDiscoveryId` (need to change the BlobAPI code to ensure the discovery key is used)
- Having both `hash` and `name` feels redundant, since they are both unique identifiers. However thinking about it more, if in the future we allow users to attach files from outside Mapeo, it is possible that a user could attach the same file twice (e.g. same hash), but maybe we should just check for that and return the ID without writing a copy. That can probably be done in a later release if we support external attachments. So, TL;DR, I think we only need the hash. We can call it `name` and say it is the sha256 hash, or we can remove the name property and call it `sha256` or `hash`. Any preferences? ... However, thinking about this more, keeping the name short (e.g. random 8-bytes) vs. the 32-byte sha does keep the size of the `blobIndex` core down (3 variants * 24 bytes less per index record, or 3 * 48 bytes since will be hex-encoded string = 144 bytes) at the expense of adding 8 extra bytes to the `data` core (in the attachment record) for the name, so maybe... keep them both. However reading a blob should not require the hash, but the returned value from creating a blob should include the hash. Clear as mud?
- `variant` shouldn't be on the `Attachment` type, because the attachment refers to all variants.

**Comment by achou11** [comment]
> driveId for the BlobId should be changed to driveDiscoveryId (need to change the BlobAPI code to ensure the discovery key is used)

Pretty sure this is already doing what you're suggesting, and it's just a matter of confusing naming

https://github.com/digidem/mapeo-core-next/blob/1f4abac31b84fcc5a73a9db2ef7f8fafbffeb3ba/src/blob-store/index.js#L74

https://github.com/digidem/mapeo-core-next/pull/294#pullrequestreview-1655015372 (lol)

> Clear as mud?

i'm...not sure what i just read 😂

**Comment by gmaclennan** [comment]
To summarize my jumbled comments above:
- `type` - no change necessary. Front-end should handle `UNRECOGNIZED` on read, but it is not valid for writes.
- `BlobId.driveId` is fine to leave for now, it is the discovery ID, just needs some renaming along with many other things for consistency at a later stage.
- Keep both `hash` and `name` in the attachment record. Use `name` for lookup. `hash` is just used for verification. Other props for verification may be added to attachment records in the future.
- `variant` is not on the attachment record because each attachment could have more than one variant. In the future we might want to populate this field at read-time to append the attachment variants that are actually available.
TL;DR no action required, closing.
